service/storage_service.cc

```diff
@@ -1149,6 +1172,133 @@ class topology_coordinator {
+            auto s = _db.find_schema(table);
+            for (auto&& [tablet, trinfo] : tmap.transitions()) {
+                co_await coroutine::maybe_yield();
+                auto last_token = tmap.get_last_token(tablet);
+                out.emplace_back(
+                    replica::tablet_mutation_builder(guard.write_timestamp(), s->ks_name(), table)
+                        .set_stage(last_token, stage)
+                        .build());
+            }
+        }
+    }
+
+    future<> del_tablet_transitions(std::vector<canonical_mutation>& out, group0_guard& guard) {
+        auto tm = get_token_metadata_ptr();
+        for (auto&& [table, tmap] : tm->tablets().all_tables()) {
+            co_await coroutine::maybe_yield();
+            auto s = _db.find_schema(table);
+            for (auto&& [tablet, trinfo] : tmap.transitions()) {
+                co_await coroutine::maybe_yield();
+                auto last_token = tmap.get_last_token(tablet);
+                out.emplace_back(
+                    replica::tablet_mutation_builder(guard.write_timestamp(), s->ks_name(), table)
+                        .del_transition(last_token)
+                        .set_replicas(last_token, trinfo.next)
+                        .build());
+            }
+        }
+    }
+
+    future<> transition_tablets_to_stage(group0_guard guard, locator::tablet_transition_stage stage, topology::transition_state topo_state) {
+        std::vector<canonical_mutation> updates;
+        co_await set_tablet_transition_stage(updates, guard, stage);
+        updates.emplace_back(
+            topology_mutation_builder(guard.write_timestamp())
+                .set_transition_state(topo_state)
+                .set_version(_topo_sm._topology.version + 1)
+                .build());
+        co_await update_topology_state(std::move(guard), std::move(updates), format("Moved tablet migration to stage: {}", stage));
+    }
+
+    future<> handle_tablet_allow_write_both_read_old(group0_guard guard) {
+        guard = co_await global_tablet_token_metadata_barrier(std::move(guard));
+        co_await transition_tablets_to_stage(std::move(guard),
+                                     locator::tablet_transition_stage::write_both_read_old,
+                                     topology::transition_state::tablet_write_both_read_old);
+    }
+
+    future<> handle_tablet_write_both_read_old(group0_guard guard) {
+        guard = co_await global_tablet_token_metadata_barrier(std::move(guard));
+        co_await transition_tablets_to_stage(std::move(guard),
+                                     locator::tablet_transition_stage::streaming,
+                                     topology::transition_state::tablet_streaming);
+    }
+
+    // The state "streaming" is needed to ensure that stale stream_tablet() RPC doesn't
+    // get admitted before global_tablet_token_metadata_barrier() is finished in
+    // handle_tablet_write_both_read_old().
```

Comment:
Can you elaborate. I am not sure I understand how having separate "streaming" stage guards from stale RPCs. Wouldn't sending streaming command through `raft_topology_cmd` rpc solve this without the state?

Reply:
I cannot reuse current raft_topology_cmd because it doesn't support parallelism

Reply:
> Having "streaming" stage means there is no work for the coordinator other than triggering streaming, so as soon as the relica sees "streaming" stage it means it's ok to trigger streaming and we don't need to worry about guarding against old rpcs.

What prevents very old RPC to arrive when replica is in the "streaming" stage. I guess I am missing something here.

Reply:
> I cannot reuse current raft_topology_cmd because it doesn't support parallelism

In what way it does not support it? You cannot call `raft_topology_cmd` multiple times on the same node simultaneously by different RPC (because of `_raft_topology_cmd_handler_state presumably)? But is it really needed? Can we pass all the work to a node in one RPC? Will be more efficient as well.

Reply:
> What prevents very old RPC to arrive when replica is in the "streaming" stage. I guess I am missing something here.

Nothing prevents that, but in this implementation it's ok if the old RPC does trigger streaming, because it will do what the coordinator wants to do right now anyway. The new RPC will block on that streaming (retry path).

Reply:
> In what way it does not support it? You cannot call `raft_topology_cmd` multiple times on the same node simultaneously by different RPC (because of `_raft_topology_cmd_handler_state presumably)? But is it really needed? Can we pass all the work to a node in one RPC? Will be more efficient as well.

Yes, we can have only one raft_topology_cmd per node at a time due to the check for cmd_index. If we submitted them in parallel, one of them could get reordered and get rejected. We could batch requests per node in this PR, because we stream in tandem, but this will change in the next PR, we want to stream tablets independently, and may start new sessions while old sessions are still running.

Reply:
> Nothing prevents that, but in this implementation it's ok if the old RPC does trigger streaming, because it will do what the coordinator wants to do right now anyway. The new RPC will block on that streaming (retry path).

So why do you prefer to have additional "streaming" state instead of filtering out outdated RPCs?

Reply:
> Yes, we can have only one raft_topology_cmd per node at a time due to the check for cmd_index. If we submitted them in parallel, one of them could get reordered and get rejected. We could batch requests per node in this PR, because we stream in tandem, but this will change in the next PR, we want to stream tablets independently, and may start new sessions while old sessions are still running.

Would adding a lock around cmd_index help?

Reply:
> So why do you prefer to have additional "streaming" state instead of filtering out outdated RPCs?

To avoid maintaining extra per-tablet state on each replica. Adding a stage seems simpler.

Reply:
> > So why do you prefer to have additional "streaming" state instead of filtering out outdated RPCs?
> 
> To avoid maintaining extra per-tablet state on each replica. Adding a stage seems simpler.

Why should it be per-tablet as opposite to global like now?

Reply:
> Would adding a lock around cmd_index help?

I don't immediately see how. Suppose you start streaming for tablet X, and immediately after but in in the same batch, start streaming for Y. RPCs for X and Y will race on the receiving replica. They may get reordered while executing the read barrier. Unless we somehow prevent reordering in all the layers involved, but that seems fragile.

Reply:
> Why should it be per-tablet as opposite to global like now?

Because of the same problem with parallelism and reordering we talk about regarding raft_topology_cmd.

Reply:
> I don't immediately see how. Suppose you start streaming for tablet X, and immediately after but not in the same batch, start streaming for Y. RPCs for X and Y will race on the receiving replica. They may get reordered while executing the read barrier. Unless we somehow prevent reordering in all the layers involved, but that seems fragile.

It can be solved by a lock on rpc handler level. On RPC level itself reordering may only happen in case of a failed connection which is rare and can be handled by simply retrying an rpc on the coordinator. But I see what you mean. It is unfortunate though that outdated RPCs can be executed. Much harder to reason about things.